<div style="background: #000;
            color: #FFF;
            margin: 0px;
            padding: 10px 0px 20px 0px;
            text-align: center; 
                ">
    <h1>Data: Types, Integrity and Accuracy</h1>
</div>

### Numerical
**Numerical data** is information that is measurable, and it is, of course, data represented as numbers and not words or text.

**Continuous numbers** are numbers that don’t have a logical end to them. Examples include variables that represent money or height.

**Discrete numbers** are the opposite; they have a logical end to them. Some examples include variables for days in the month, or number of bugs logged.

# Categorical  

For categorical data, this is any data that isn’t a number, which can mean a string of text or date. These variables can be broken down into nominal and ordinal values, though you won’t often see this done.

**Ordinal values** are values that have a set order to them. Examples of ordinal values include having a priority on a bug such as “Critical” or “Low” or the ranking of a race as “First” or “Third”. Nominal values are the opposite of ordinal values, and they represent values with no set order to them. Nominal value examples include variables such as “Country” or “Marital Status”.

In addition to ordinal and nominal values, there is a special type of categorical data called **binary**. Binary data types only have two values – yes or no. This can be represented in different ways such as “True” and “False” or 1 and 0. Binary data is used heavily for classification machine learning models. Examples of binary variables can include whether a person has stopped their subscription service or not, or if a person bought a car or not.

### Where to get data?

There are lots of way to get data, and lots of places to get it from. Typically, most of this data will be accessed through the internet, in one way or another, especially when pursuing indepent research projects.
Institutional Access

If you are working with data as part of an institution, such as a company of research lab, the institution will typically have data it needs analyzing, that it collects in various ways. Keep in mind that even people working inside institutions, with access to local data, will data still seek to find and incorporate external datasets.
Data Repositories

Data repositories are databases from which you can download data. Some data repositories allow you to explore available datasets and download datasets in bulk. Others may also offer APIs, through which you can request specific data from particular databases.
Web Scraping

The web itself is full of unstructured data. Web scraping can be done to directly extract and collect data directly from websites.
Asking People for Data

Not all data is indexed or accessible on the web, at least not publicly. Sometimes finding data means figuring out if any data is available, figuring out where it might be, and then reaching out and asking people directly about data access. If there is some particular data you need, you can try to figure out who might have it, and get in touch to see if it might be available.
Data Gathering Skills

Depending on your gathering method, you will likely have to do some combination of the following:
* Direct download data files from repositories
* Query databases & use APIs to extract and collect data of interest
* Ask people for data, and going to pick up data with a harddrive

Ultimately, the goal is collect and curate data files, hopefully structured, that you can read into Python.


### Structured and Unstructured data

**Structured Data**

Structured data is the data which conforms to a data model, has a well define structure, follows a consistent order and can be easily accessed and used by a person or a computer program.

Structured data is usually stored in well-defined schemas such as Databases. It is generally tabular with column and rows that clearly define its attributes.

SQL (Structured Query language) is often used to manage structured data stored in databases.

Characteristics of Structured Data:
* Data conforms to a data model and has easily identifiable structure
* Data is stored in the form of rows and columns (example : Database)
* Data is well organised so, Definition, Format and Meaning of data is explicitly known
* Data resides in fixed fields within a record or file
* Similar entities are grouped together to form relations or classes
* Entities in the same group have same attributes
* Easy to access and query, So data can be easily used by other programs
* Data elements are addressable, so efficient to analyse and process

**Unstructured Data**

Unstructured data is everywhere. In fact, most individuals and organizations conduct their lives around unstructured data. Just as with structured data, unstructured data is either machine generated or human generated.

Here are some examples of machine-generated unstructured data:
* Satellite images: This includes weather data or the data that the government captures in its satellite surveillance imagery. Just think about Google Earth, and you get the picture.
* Scientific data: This includes seismic imagery, atmospheric data, and high energy physics.
* Photographs and video: This includes security, surveillance, and traffic video.
* Radar or sonar data: This includes vehicular, meteorological, and oceanographic seismic profiles.

The following list shows a few examples of human-generated unstructured data:
* Text internal to your company: Think of all the text within documents, logs, survey results, and e-mails. Enterprise information actually represents a large percent of the text information in the world today.
* Social media data: This data is generated from the social media platforms such as YouTube, Facebook, Twitter, LinkedIn, and Flickr.
* Mobile data: This includes data such as text messages and location information.
* website content: This comes from any site delivering unstructured content, like YouTube, Flickr, or Instagram.

### Why does unstructured data matter?

**Because 80% of the data in companies is unstructured**, organizations need to understand the types of unstructured data they are accumulating and the best ways to process and store this data for business advantages. Without data management strategies and guidance in these areas, companies run the risks of not capitalizing on unstructured data, failing to keep up with competitors, or storing more unstructured data than they really need, thereby running up data center costs.

In a majority of cases, unstructured data is ultimately related back to the company's structured data records. As an example, every x-ray or MRI image for a patient is related back to the patient's record in the hospital's record system. The patient record in the record system is enriched with unstructured data that is linked to it, and the doctor gets a more complete picture of the patient.

This is the value of unstructured data: It enriches corporate data and enables leaders to work smarter.

### Ways you can deal with unstructured data:
[source](https://www.ironmountain.com/blogs/2020/4-ways-to-deal-with-unstructured-data)

1. Throw It Away

The reality is that much of the data organizations collect isn’t very interesting or useful, but it still takes up a lot of storage space. Devices such as smart cameras and machine sensors create huge amounts of data, little of which is needed if everything is operating normally. Rather than storing all that data, a better solution is evaluating it before it hits the network and discarding what isn’t needed. Edge computing, a type of distributed processing that makes decisions close to where data is gathered, is a promising way to do just that.

Edge AI is a special category of artificial intelligence that’s specifically intended to make decisions requiring immediate attention, such as controlling the brakes of a car or determining the likelihood that a machine is about to fail, according to Forbes. Edge AI can also be used to scour data streams and quickly identify what:
* Can be discarded.
* Requires immediate attention.
* Should be stored for analysis.

There are high hopes for edge computing; many researchers expect the market to grow more than 30% annually for the next several years, according to Allied Market Research.

2. Deduplicate It

Have you ever been on the distribution list of a mass email that included a 15-megabyte PowerPoint attachment? Organizations generate an enormous amount of duplicate data, so much so that a 2013 IDC study estimated that companies spend $44 billion annually to store copies of data they already have. This situation has likely only grown since then.

High-speed, in-line deduplication can flag duplicate records and either hold them for review or delete them automatically. While savings vary when using this method, it led to more than a 10 times reduction in storage space for two-thirds of the companies in an Enterprise Strategy Group survey shared by TechTarget.

3. Tier It

If your organization treats all data the same, it’s flushing money down the drain. Only a tiny percentage of data is typically mission-critical enough that it needs to be instantly available on expensive storage media. Most data can be relegated to spinning disks or tape. Tiered storage automatically assigns data to the most appropriate storage medium based upon policies, which often results in significant savings.

Horison Information Strategies estimates that between 63% and 85% of a typical organization’s data can be moved to secondary or long-term storage without impacting operations. If your organization isn’t moving infrequently accessed data from disk to tape or from cloud to tape, you should take a fresh look at this low-cost archival option.

Tape is the most cost-effective storage medium and its retrieval times are approaching that of disk, thanks to recent advances in technology such as redundant arrays of independent tape. What’s more, tape is stored offline, making it nearly impervious to malware attacks. Additionally, cloud services can archive data to tape automatically, which can help you save big while still retaining ready access to your data.

4. Structure It

Machine learning algorithms are great at finding patterns, including those in unstructured data sources such as text documents and images. By repeatedly scanning similar documents with some human oversight, machines can quickly figure out that a sequence of digits is more likely to be a Social Security number than a phone number or the identities of people in a photo or video. This semi-structured data can then be loaded into databases for analytical processing.

When deciding how to deal with unstructured data, look to cloud and tape for ideas. Deduplicating data helps minimize how much data is stored and AI helps with information processing and analysis, but cloud to tape storage ensures your data is protected in the most secure, accessible and cost-effective manner possible.

### Dirty Data

Dirty data, also known as rogue data, are inaccurate, incomplete or inconsistent data, especially in a computer system or database.

Dirty data can contain such mistakes as spelling or punctuation errors, incorrect data associated with a field, incomplete or outdated data, or even data that has been duplicated in the database. They can be cleaned through a process known as data cleansing.

### Data cleansing

Data cleansing or data cleaning is the process of detecting and correcting (or removing) corrupt or inaccurate records from a record set, table, or database and refers to identifying incomplete, incorrect, inaccurate or irrelevant parts of the data and then replacing, modifying, or deleting the dirty or coarse data. Data cleansing may be performed interactively with data wrangling tools, or as batch processing through scripting.

After cleansing, a data set should be consistent with other similar data sets in the system. The inconsistencies detected or removed may have been originally caused by user entry errors, by corruption in transmission or storage, or by different data dictionary definitions of similar entities in different stores. Data cleaning differs from data validation in that validation almost invariably means data is rejected from the system at entry and is performed at the time of entry, rather than on batches of data. 

### Requisites of quality data

High-quality data needs to pass a set of quality criteria. Those include:
* **Validity**: The degree to which the measures conform to defined business rules or constraints (see also Validity (statistics)). When modern database technology is used to design data-capture systems, validity is fairly easy to ensure: invalid data arises mainly in legacy contexts (where constraints were not implemented in software) or where inappropriate data-capture technology was used (e.g., spreadsheets, where it is very hard to limit what a user chooses to enter into a cell, if cell validation is not used). Data constraints fall into the following categories:
  + Data-Type Constraints – e.g., values in a particular column must be of a particular data type, e.g., Boolean, numeric (integer or real), date, etc.
  + Range Constraints: typically, numbers or dates should fall within a certain range. That is, they have minimum and/or maximum permissible values.
  + Mandatory Constraints: Certain columns cannot be empty.
  +  Unique Constraints: A field, or a combination of fields, must be unique across a dataset. For example, no two persons can have the same social security number.
  + Set-Membership constraints: The values for a column come from a set of discrete values or codes. For example, a person's gender may be Female, Male or Unknown (not recorded).
  + Foreign-key constraints: This is the more general case of set membership. The set of values in a column is defined in a column of another table that contains unique values. For example, in a US taxpayer database, the "state" column is required to belong to one of the US's defined states or territories: the set of permissible states/territories is recorded in a separate State table. The term foreign key is borrowed from relational database terminology.
  + Regular expression patterns: Occasionally, text fields will have to be validated this way. For example, phone numbers may be required to have the pattern (999) 999-9999.
  + Cross-field validation: Certain conditions that utilize multiple fields must hold. For example, in laboratory medicine, the sum of the components of the differential white blood cell count must be equal to 100 (since they are all percentages). In a hospital database, a patient's date of discharge from the hospital cannot be earlier than the date of admission.
* **Accuracy**: The degree of conformity of a measure to a standard or a true value - see also Accuracy and precision. Accuracy is very hard to achieve through data-cleansing in the general case because it requires accessing an external source of data that contains the true value: such "gold standard" data is often unavailable. Accuracy has been achieved in some cleansing contexts, notably customer contact data, by using external databases that match up zip codes to geographical locations (city and state) and also help verify that street addresses within these zip codes actually exist.
* **Completeness**: The degree to which all required measures are known. Incompleteness is almost impossible to fix with data cleansing methodology: one cannot infer facts that were not captured when the data in question was initially recorded. (In some contexts, e.g., interview data, it may be possible to fix incompleteness by going back to the original source of data, i.e. re-interviewing the subject, but even this does not guarantee success because of problems of recall - e.g., in an interview to gather data on food consumption, no one is likely to remember exactly what one ate six months ago. In the case of systems that insist certain columns should not be empty, one may work around the problem by designating a value that indicates "unknown" or "missing", but the supplying of default values does not imply that the data has been made complete.)
* **Consistency**: The degree to which a set of measures are equivalent in across systems (see also Consistency). Inconsistency occurs when two data items in the data set contradict each other: e.g., a customer is recorded in two different systems as having two different current addresses, and only one of them can be correct. Fixing inconsistency is not always possible: it requires a variety of strategies - e.g., deciding which data were recorded more recently, which data source is likely to be most reliable (the latter knowledge may be specific to a given organization), or simply trying to find the truth by testing both data items (e.g., calling up the customer).
* **Uniformity**: The degree to which a set data measures are specified using the same units of measure in all systems ( see also Unit of measure). In datasets pooled from different locales, weight may be recorded either in pounds or kilos and must be converted to a single measure using an arithmetic transformation.

The term **integrity** encompasses accuracy, consistency and some aspects of validation  but is rarely used by itself in data-cleansing contexts because it is insufficiently specific. (For example, "referential integrity" is a term used to refer to the enforcement of foreign-key constraints above.) 


### Data Integrity

Data integrity is the maintenance of, and the assurance of, data accuracy and consistency over its entire life-cycle and is a critical aspect to the design, implementation, and usage of any system that stores, processes, or retrieves data. The term is broad in scope and may have widely different meanings depending on the specific context – even under the same general umbrella of computing. It is at times used as a proxy term for data quality, while data validation is a pre-requisite for data integrity. Data integrity is the opposite of data corruption. The overall intent of any data integrity technique is the same: ensure data is recorded exactly as intended (such as a database correctly rejecting mutually exclusive possibilities). Moreover, upon later retrieval, ensure the data is the same as when it was originally recorded. In short, data integrity aims to prevent unintentional changes to information. Data integrity is not to be confused with data security, the discipline of protecting data from unauthorized parties. 

<div style="background: #000;
            color: #FFF;
            margin: 0px;
            padding: 10px 0px 20px 0px;
            text-align: center; 
                ">
    <h1>Code Along</h1>
    <h3>Cleaning up dirty data</h3>
</div>

Some new things we'll cover in this code along:

[pandas.DataFrame.apply](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.apply.html) - applies a function to rows of a col  
[pandas.DataFrame.drop](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) - drops rows or columns  
[pandas.DataFrame.drop_duplicates](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) - drops duplicates  
[pandas.DataFrame.duplicated](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) - returns True if duplicate  
[pandas.DataFrame.fillna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html) - fills NaN values  
[pandas.merge](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) - merges two dfs  

In [43]:
import numpy as np 
import pandas as pd 
import datetime

Let's open the `ted_talks.csv` file and generate a dataframe. Then we'll check the top of the `df` to see what we have.

In [ ]:
df = pd.read_csv("ted_talks.csv")
df.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869

Checking the shape of the `df`

In [45]:
df.shape

(2550, 17)

Checking info to see other information (like how much RAM our dataset uses and how many non-null entries we have)

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2550 entries, 0 to 2549
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   comments            2550 non-null   int64 
 1   description         2550 non-null   object
 2   duration            2550 non-null   int64 
 3   event               2550 non-null   object
 4   film_date           2550 non-null   int64 
 5   languages           2550 non-null   int64 
 6   main_speaker        2531 non-null   object
 7   name                2550 non-null   object
 8   num_speaker         2550 non-null   int64 
 9   published_date      2550 non-null   int64 
 10  ratings             2550 non-null   object
 11  related_talks       2550 non-null   object
 12  speaker_occupation  2544 non-null   object
 13  tags                2550 non-null   object
 14  title               2550 non-null   object
 15  url                 2550 non-null   object
 16  views               2550

We see we have 2555 rows but 2531 non-null rows for 'main_speaker'. So let's check that out.

In [47]:
df[df['main_speaker'].isna()][["description", "main_speaker","name"]]

description main_speaker  \
255   Nine days before TED2008, filmmaker David Hoff...          NaN   
641   Games like World of Warcraft give players the ...          NaN   
815   Join John Hardy on a tour of the Green School,...          NaN   
845   Technology is evolving us, says Amber Case, as...          NaN   
891   Lighting architect Rogier van der Heide offers...          NaN   
951   Designer Jessi Arrington packed nothing for TE...          NaN   
953   Renowned paleontologist Jack Horner has spent ...          NaN   
1250  Bartenders need to know your age, retailers ne...          NaN   
1285  Designer Jared Ficklin creates wild visualizat...          NaN   
1479  Today, thanks to better early detection, there...          NaN   
1626  Ecologist Lian Pin Koh makes a persuasive case...          NaN   
1633  The one thing all humans have in common is tha...          NaN   
1648  It's amazing what a rigged game of Monopoly ca...          NaN   
1794  Tour the deep dark world of the East German st...          NaN   
1973  Singer-songwriter Dawn Landes tells the story ...          NaN   
1998  Decades ago, few pediatricians had heard of au...          NaN   
2133  How much do you know about intellectual disabi...          NaN   
2185  "Copy editing for The New Yorker is like playi...          NaN   
2196  Secrets, disease and beauty are all written in...          NaN   

                                                   name  
255   David Hoffman: What happens when you lose ever...  
641      Jane McGonigal: Gaming can make a better world  
815                   John Hardy: My green school dream  
845                  Amber Case: We are all cyborgs now  
891      Rogier van der Heide: Why light needs darkness  
951                Jessi Arrington: Wearing nothing new  
953     Jack Horner: Building a dinosaur from a chicken  
1250      David Birch: A new way to stop identity theft  
1285  Jared Ficklin: New ways to see music (with col...  
1479  Thomas Insel: Toward a new understanding of me...  
1626   Lian Pin Koh: A drone's-eye view of conservation  
1633  David Steindl-Rast: Want to be happy? Be grateful  
1648               Paul Piff: Does money make you mean?  
1794  Hubertus Knabe: The dark secrets of a surveill...  
1973  Dawn Landes: A song for my hero, the woman who...  
1998   Steve Silberman: The forgotten history of autism  
2133  Matthew Williams: Special Olympics let me be m...  
2185  Mary Norris: The nit-picking glory of The New ...  
2196  Riccardo Sabatini: How to read the genome and ...

Before we fix the "main_speaker" column, let's shrink our dataset by removing duplicates. 
First we need to see how many duplicates we have:

In [48]:
df['url'].duplicated().sort_values(ascending=False)[:6]

1592    True
2197    True
2079    True
448     True
2349    True
2428    True
Name: url, dtype: bool

In [49]:
df[(df['url'].duplicated() == False)]

comments                                        description  duration  \
0         4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1          265  With the same humor and humanity he exuded in ...       977   
2          124  New York Times columnist David Pogue takes aim...      1286   
3          200  In an emotionally charged talk, MacArthur-winn...      1116   
4          593  You've never seen data presented like this. Wi...      1190   
...        ...                                                ...       ...   
2545        17  Between 2008 and 2016, the United States depor...       476   
2546         6  How can you study Mars without a spaceship? He...       290   
2547        10  Science fiction visions of the future show us ...       651   
2548        32  In an unmissable talk about race and politics ...      1100   
2549         8  With more than half of the world population li...       519   

             event   film_date  languages         main_speaker  \
0          TED2006  1140825600         60         Ken Robinson   
1          TED2006  1140825600         43              Al Gore   
2          TED2006  1140739200         26          David Pogue   
3          TED2006  1140912000         35        Majora Carter   
4          TED2006  1140566400         48         Hans Rosling   
...            ...         ...        ...                  ...   
2545       TED2017  1496707200          4     Duarte Geraldino   
2546       TED2017  1492992000          3  Armando Azua-Bustos   
2547       TED2017  1492992000          1       Radhika Nagpal   
2548  TEDxMileHigh  1499472000          1     Theo E.J. Wilson   
2549       TED2017  1492992000          1    Karoliina Korppoo   

                                                   name  num_speaker  \
0             Ken Robinson: Do schools kill creativity?            1   
1                  Al Gore: Averting the climate crisis            1   
2                         David Pogue: Simplicity sells            1   
3                    Majora Carter: Greening the ghetto            1   
4         Hans Rosling: The best stats you've ever seen            1   
...                                                 ...          ...   
2545  Duarte Geraldino: What we're missing in the de...            1   
2546  Armando Azua-Bustos: The most Martian place on...            1   
2547  Radhika Nagpal: What intelligent machines can ...            1   
2548  Theo E.J. Wilson: A black man goes undercover ...            1   
2549  Karoliina Korppoo: How a video game might help...            1   

      published_date                                            ratings  \
0         1151367060  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1         1151367060  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2         1151367060  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3         1151367060  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4         1151440680  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   
...              ...                                                ...   
2545      1505851216  [{'id': 3, 'name': 'Courageous', 'count': 24},...   
2546      1505919737  [{'id': 22, 'name': 'Fascinating', 'count': 32...   
2547      1506006095  [{'id': 1, 'name': 'Beautiful', 'count': 14}, ...   
2548      1506024042  [{'id': 11, 'name': 'Longwinded', 'count': 3},...   
2549      1506092422  [{'id': 21, 'name': 'Unconvincing', 'count': 2...   

                                          related_talks  \
0     [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1     [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2     [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3     [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4     [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   
...                                                 ...   
2545  [{'id': 2596, 'hero': 'https://pe.tedcdn.com/i...   
2546  [{'id': 2491, 'hero'

In [50]:
df[df['url'].duplicated()]

comments                                        description  duration  \
448        133  Nandan Nilekani, the visionary co-founder of o...       919   
546         72  The world turns on symmetry -- from the spin o...      1099   
1022       231  Modern medicine is in danger of losing a power...      1112   
1281       554  When game designer Jane McGonigal found hersel...      1170   
1520       103  For 28 years, Denise Herzing has spent five mo...       878   
1592       404  Amy Webb was having no luck with online dating...      1047   
1605       165  How do we decide who we are? Hetain Patel's su...       546   
1651       683  "I have cerebral palsy. I shake all the time,"...       853   
1811       184  Which of the following is awesome: your lunch ...       373   
1833       271  Brain imaging pioneer Nancy Kanwisher, who use...      1060   
1835       120  French economist Thomas Piketty caused a sensa...      1260   
1877        48  Choreographer Aakash Odedra is dyslexic and ha...       590   
2079        56  Although the tapir is one of the world's large...       692   
2197        24  After Sarah Gray's unborn son Thomas was diagn...       617   
2266        35  The destruction of war doesn't stop when the f...       696   
2349        35  With words like shards of glass, Chinaka Hodge...       237   
2428        11  If Yoda goes into cardiac arrest, will you kno...       322   
2451        70  How can we harness the power of superintellige...      1055   
2516        18  "Music is everywhere, and it is in everything,...       558   

                     event   film_date  languages      main_speaker  \
448                TED2009  1233705600         24   Nandan Nilekani   
546         TEDGlobal 2009  1248393600         23  Marcus du Sautoy   
1022        TEDGlobal 2011  1310601600         33  Abraham Verghese   
1281        TEDGlobal 2012  1340841600         30    Jane McGonigal   
1520               TED2013  1361923200         26    Denise Herzing   
1592       TEDSalon NY2013  1365638400         32          Amy Webb   
1605        TEDGlobal 2013  1370908800         36      Hetain Patel   
1651         TEDWomen 2013  1386115200         42     Maysoon Zayid   
1811               TED@NYC  1404864000         30      Jill Shargaa   
1833               TED2014  1394755200         28   Nancy Kanwisher   
1835  TEDSalon Berlin 2014  1403481600         24    Thomas Piketty   
1877        TEDGlobal 2014  1413849600         39     Aakash Odedra   
2079      TED Fellows 2015  1440547200         28   Patrícia Medici   
2197           TEDMED 2015  1447804800         30        Sarah Gray   
2266             TEDSummit  1466985600         23    Laura Boushnak   
2349         TEDWomen 2016  1477612800         26     Chinaka Hodge   
2428                TEDNYC  1485388800         20        Todd Scott   
2451               TED2017  1492992000         10    Stuart Russell   
2516               TED2017  1492992000          7     Anika Paulson   

                                                   name  num_speaker  \
448           Nandan Nilekani: Ideas for India's future            1   
546        Marcus du Sautoy: Symmetry, reality's riddle            1   
1022                 Abraham Verghese: A doctor's touch            1   
1281  Jane McGonigal: The game that can give you 10 ...            1   
1520  Denise Herzing: Could we speak the language of...            1   
1592               Amy Webb: How I hacked online dating            1   
1605                Hetain Patel: Who am I? Think again            2   
1651  Maysoon Zayid: I got 99 problems ... palsy is ...            1   
1811  Jill Shargaa: Please, please, people. Let's pu...            1   
1833  Nancy Kanwisher: A neural portrait of the huma...            1   
1835  Thomas Piketty: New thoughts on capital in the...            1   
1877  Aakash Odedra: A dance in a hurricane of paper...            1   
2079  Patrícia Medici: The coolest animal you know n...            1

Then we'll drop the duplicates using the `df.drop_duplicates` method. 

Remember, the keyword `inplace` means it will mutate the original `df`.

In [51]:
help(df.drop_duplicates)

Help on method drop_duplicates in module pandas.core.frame:

drop_duplicates(subset: 'Hashable | Sequence[Hashable] | None' = None, *, keep: 'DropKeep' = 'first', inplace: 'bool' = False, ignore_index: 'bool' = False) -> 'DataFrame | None' method of pandas.core.frame.DataFrame instance
    Return DataFrame with duplicate rows removed.

    Considering certain columns is optional. Indexes, including time indexes
    are ignored.

    Parameters
    ----------
    subset : column label or sequence of labels, optional
        Only consider certain columns for identifying duplicates, by
        default use all of the columns.
    keep : {'first', 'last', ``False``}, default 'first'
        Determines which duplicates (if any) to keep.

        - 'first' : Drop duplicates except for the first occurrence.
        - 'last' : Drop duplicates except for the last occurrence.
        - ``False`` : Drop all duplicates.

    inplace : bool, default ``False``
        Whether to modify the DataFrame 

In [52]:
df.drop_duplicates('url', inplace=True)

If we look at `published_date` and `film_date`, they're formatted as unix timestamps. We might need a human-readable format.

[more on unix timestamps](https://en.wikipedia.org/wiki/Unix_time)

In [53]:
df[["published_date", "film_date"]]

published_date   film_date
0         1151367060  1140825600
1         1151367060  1140825600
2         1151367060  1140739200
3         1151367060  1140912000
4         1151440680  1140566400
...              ...         ...
2545      1505851216  1496707200
2546      1505919737  1492992000
2547      1506006095  1492992000
2548      1506024042  1499472000
2549      1506092422  1492992000

[2531 rows x 2 columns]

Let's see how we can turn a unix timestamp into a human-readable format

In [54]:
ts = 1151367060

s = datetime.datetime.fromtimestamp(ts)

s.strftime("%m/%d/%Y, %H:%M:%S")

'06/26/2006, 20:11:00'

Putting that together, we can create a function that does this for any timestamp

In [55]:
def transform_date(timestamp):
    _date = datetime.datetime.fromtimestamp(timestamp)
    return _date.strftime("%m/%d/%Y, %H:%M:%S")

And then we can test it with a random timestamp

In [56]:
transform_date(1151367060)

'06/26/2006, 20:11:00'

We can use the `df.apply` method to apply that function to each row of a specific column.

In [57]:
df['film_date'].apply(transform_date)

0       02/24/2006, 19:00:00
1       02/24/2006, 19:00:00
2       02/23/2006, 19:00:00
3       02/25/2006, 19:00:00
4       02/21/2006, 19:00:00
                ...         
2545    06/05/2017, 20:00:00
2546    04/23/2017, 20:00:00
2547    04/23/2017, 20:00:00
2548    07/07/2017, 20:00:00
2549    04/23/2017, 20:00:00
Name: film_date, Length: 2531, dtype: object

Seeing that it works, we can now use that `transform_date` method along with `apply` to create two new columns of human-readable timestamps

In [58]:
df['film_date_hr'] = df['film_date'].apply(transform_date)
df['published_date_hr'] = df['published_date'].apply(transform_date)

If we wanted to change the ordering of the columns, we can use `df.columns.tolist()` to list our columns and then `df.reindex()` to reindex the columns by passing in a list of columns in the order we want

In [59]:
df.columns.tolist()

['comments',
 'description',
 'duration',
 'event',
 'film_date',
 'languages',
 'main_speaker',
 'name',
 'num_speaker',
 'published_date',
 'ratings',
 'related_talks',
 'speaker_occupation',
 'tags',
 'title',
 'url',
 'views',
 'film_date_hr',
 'published_date_hr']

In [60]:
df = df.reindex(['comments',
 'description',
 'duration',
 'event',
 'film_date',
 'film_date_hr',
 'languages',
 'main_speaker',
 'name',
 'num_speaker',
 'published_date',
 'published_date_hr',
 'ratings',
 'related_talks',
 'speaker_occupation',
 'tags',
 'title',
 'url',
 'views',
],axis=1)

We can see that works by checking the head of the `df`

In [61]:
df.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date          film_date_hr  languages   main_speaker  \
0  TED2006  1140825600  02/24/2006, 19:00:00         60   Ken Robinson   
1  TED2006  1140825600  02/24/2006, 19:00:00         43        Al Gore   
2  TED2006  1140739200  02/23/2006, 19:00:00         26    David Pogue   
3  TED2006  1140912000  02/25/2006, 19:00:00         35  Majora Carter   
4  TED2006  1140566400  02/21/2006, 19:00:00         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

      published_date_hr                                            ratings  \
0  06/26/2006, 20:11:00  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  06/26/2006, 20:11:00  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  06/26/2006, 20:11:00  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  06/26/2006, 20:11:00  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  06/27/2006, 16:38:00  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869

Next, we'll fix the durations, which are in seconds, by converting them to mins and rounding the result to 2 places after the decimal

In [62]:
df['duration'] = round( df['duration'] / 60, 2)

We'll fix incorrect values for `languages` and `num_speakers` by replacing erroneous values with `1` since we know there's at least one language translated (english) and there's at least one speaker.

In [63]:
df[df['languages'] == 0]

comments                                        description  duration  \
58         222  Two Pilobolus dancers perform "Symbiosis." Doe...     13.75   
115         27  The avant-garde string quartet Ethel performs ...      3.57   
135         36  After Vusi Mahlasela's 3-song set at TEDGlobal...      4.98   
209         67  Rokia Traore sings the moving "M'Bifo," accomp...      6.98   
237         43  Singer-songwriter Rokia Traore performs "Kouna...      6.43   
...        ...                                                ...       ...   
2330        71  Women's equality won't just happen -- not unle...     19.80   
2353        76  In every workplace, there are three basic kind...     13.47   
2407         5  Grammy-winning Silk Road Ensemble display thei...      6.48   
2418        11  Sō Percussion creates adventurous compositions...     10.15   
2510        50  Do you sometimes have your most creative ideas...     16.22   

               event   film_date          film_date_hr  languages  \
58           TED2005  1109289600  02/24/2005, 19:00:00          0   
115          TED2006  1138838400  02/01/2006, 19:00:00          0   
135   TEDGlobal 2007  1181260800  06/07/2007, 20:00:00          0   
209   TEDGlobal 2007  1181088000  06/05/2007, 20:00:00          0   
237   TEDGlobal 2007  1181088000  06/05/2007, 20:00:00          0   
...              ...         ...                   ...        ...   
2330   TEDWomen 2016  1477526400  10/26/2016, 20:00:00          0   
2353         TED@IBM  1479168000  11/14/2016, 19:00:00          0   
2407         TED2016  1455494400  02/14/2016, 19:00:00          0   
2418         TED2016  1455494400  02/14/2016, 19:00:00          0   
2510         TED2017  1492992000  04/23/2017, 20:00:00          0   

             main_speaker                                               name  \
58              Pilobolus                  Pilobolus: A dance of "Symbiosis"   
115                 Ethel          Ethel: A string quartet plays "Blue Room"   
135        Vusi Mahlasela                             Vusi Mahlasela: "Woza"   
209          Rokia Traore                             Rokia Traore: "M'Bifo"   
237          Rokia Traore                           Rokia Traore: "Kounandi"   
...                   ...                                                ...   
2330        Sandi Toksvig  Sandi Toksvig: A political party for women's e...   
2353           Adam Grant            Adam Grant: Are you a giver or a taker?   
2407   Silk Road Ensemble                    Silk Road Ensemble: "Turceasca"   
2418        Sō Percussion        Sō Percussion: "Music for Wood and Strings"   
2510     Manoush Zomorodi  Manoush Zomorodi: How boredom can lead to your...   

      num_speaker  published_date     published_date_hr  \
58              1      1170979860  02/08/2007, 19:11:00   
115             1      1182184140  06/18/2007, 12:29:00   
135             1      1187695440  08/21/2007, 07:24:00   
209             1      1206580680  03/26/2008, 21:18:00   
237             1      1212627600  06/04/2008, 21:00:00   
...           ...             ...                   ...   
2330           -2      1479744573  11/21/2016, 11:09:33   
2353           -1      1483457867  01/03/2017, 10:37:47   
2407            1      1489759215  03/17/2017, 10:00:15   
2418            1      1490963646  03/31/2017, 08:34:06   
2510            0      1502809223  08/15/2017, 11:00:23   

                                                ratings  \
58    [{'id': 1, 'name': 'Beautiful', 'count': 1810}...   
115   [{'id': 1, 'name': 'Beautiful', 'count': 216},...   
135   [{'id': 8, 'name': 'Informative', 'count': 4},...   
209   [{'id': 23, 'name': 'Jaw-dropping', 'count': 5...   
237   [{'id': 22, 'name': 'Fascinating', 'count': 84...   
...                                                 ...   
2330  [{'id': 3, 'name': 'Courageous', 'count': 200}...   
2353  [{'id': 3, 'name': 'Courageous', 'count': 309}...   
2407  [{'id': 1, 'name': 'B

In [64]:
df['languages'].replace(0, 1, inplace=True)
df

C:\Users\saidmf\AppData\Local\Temp\ipykernel_28780\3326432019.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['languages'].replace(0, 1, inplace=True)


comments                                        description  duration  \
0         4553  Sir Ken Robinson makes an entertaining and pro...     19.40   
1          265  With the same humor and humanity he exuded in ...     16.28   
2          124  New York Times columnist David Pogue takes aim...     21.43   
3          200  In an emotionally charged talk, MacArthur-winn...     18.60   
4          593  You've never seen data presented like this. Wi...     19.83   
...        ...                                                ...       ...   
2545        17  Between 2008 and 2016, the United States depor...      7.93   
2546         6  How can you study Mars without a spaceship? He...      4.83   
2547        10  Science fiction visions of the future show us ...     10.85   
2548        32  In an unmissable talk about race and politics ...     18.33   
2549         8  With more than half of the world population li...      8.65   

             event   film_date          film_date_hr  languages  \
0          TED2006  1140825600  02/24/2006, 19:00:00         60   
1          TED2006  1140825600  02/24/2006, 19:00:00         43   
2          TED2006  1140739200  02/23/2006, 19:00:00         26   
3          TED2006  1140912000  02/25/2006, 19:00:00         35   
4          TED2006  1140566400  02/21/2006, 19:00:00         48   
...            ...         ...                   ...        ...   
2545       TED2017  1496707200  06/05/2017, 20:00:00          4   
2546       TED2017  1492992000  04/23/2017, 20:00:00          3   
2547       TED2017  1492992000  04/23/2017, 20:00:00          1   
2548  TEDxMileHigh  1499472000  07/07/2017, 20:00:00          1   
2549       TED2017  1492992000  04/23/2017, 20:00:00          1   

             main_speaker                                               name  \
0            Ken Robinson          Ken Robinson: Do schools kill creativity?   
1                 Al Gore               Al Gore: Averting the climate crisis   
2             David Pogue                      David Pogue: Simplicity sells   
3           Majora Carter                 Majora Carter: Greening the ghetto   
4            Hans Rosling      Hans Rosling: The best stats you've ever seen   
...                   ...                                                ...   
2545     Duarte Geraldino  Duarte Geraldino: What we're missing in the de...   
2546  Armando Azua-Bustos  Armando Azua-Bustos: The most Martian place on...   
2547       Radhika Nagpal  Radhika Nagpal: What intelligent machines can ...   
2548     Theo E.J. Wilson  Theo E.J. Wilson: A black man goes undercover ...   
2549    Karoliina Korppoo  Karoliina Korppoo: How a video game might help...   

      num_speaker  published_date     published_date_hr  \
0               1      1151367060  06/26/2006, 20:11:00   
1               1      1151367060  06/26/2006, 20:11:00   
2               1      1151367060  06/26/2006, 20:11:00   
3               1      1151367060  06/26/2006, 20:11:00   
4               1      1151440680  06/27/2006, 16:38:00   
...           ...             ...                   ...   
2545            1      1505851216  09/19/2017, 16:00:16   
2546            1      1505919737  09/20/2017, 11:02:17   
2547            1      1506006095  09/21/2017, 11:01:35   
2548            1      1506024042  09/21/2017, 16:00:42   
2549            1      1506092422  09/22/2017, 11:00:22   

                                                ratings  \
0     [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1     [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2     [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3     [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4     [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   
...                                                 ...   
2545  [{'id': 3, 'name': 'Courageous', 'count': 24},...   
2546  [{'id': 22, 'name': 'Fascinating', 'count': 32...   
2547  [{'id': 1, 'name': 'Beautiful', 'count': 14},

In [65]:
df[df['languages'] == 0]

Empty DataFrame
Columns: [comments, description, duration, event, film_date, film_date_hr, languages, main_speaker, name, num_speaker, published_date, published_date_hr, ratings, related_talks, speaker_occupation, tags, title, url, views]
Index: []

In [66]:
df[(df['num_speaker'] <= 0)]

comments                                        description  duration  \
28         355  "Tipping Point" author Malcolm Gladwell gets i...     17.50   
51         499  Given $50 billion to spend, which would you so...     16.68   
66         161  Jane Goodall hasn't found the missing link, bu...     27.42   
105         99  Rural villages worldwide are being deserted, a...      3.08   
268         36  Reed Kroloff gives us a new lens for judging n...     15.35   
329         30  Graham Hawkes takes us aboard his graceful, wi...     12.18   
429         55  In this talk from TED University 2009, veteran...      4.80   
467        112  Cognitive researcher Nancy Etcoff looks at hap...     19.75   
499        210  For five years, young Emmanuel Jal fought as a...     18.05   
512        146  Vishal Vaid and his band explore a traditional...     13.57   
579        168  One of a dozen Pakistanis who came to TEDIndia...      4.47   
592        277  Kiran Bir Sethi shows how her groundbreaking R...      9.53   
681        314  We're all embedded in vast social networks of ...     20.98   
687        585  William Li presents a new way to think about t...     20.03   
719        118  Can we rebuild our broken suburbs? Ellen Dunha...     19.38   
749        285  David McCandless turns complex data sets (like...     17.93   
868         93  World champion whistler Geert Chatrou performs...     11.93   
888        226  Columnist David Brooks unpacks new insights in...     18.73   
906        154  With humor and persistence, filmmaker Morgan S...     19.47   
940        309  Terry Moore found out he'd been tying his shoe...      2.98   
976        216  After a catastrophic car accident that left hi...     22.28   
977        164  Tissue engineer and TED Fellow Nina Tandon is ...      4.22   
981        199  As political turmoil in Yemen continues, the e...     13.63   
1001       380  Using three iPods like magical props, Marco Te...      5.12   
1012       185  Despite multibillion-dollar investments in cyb...     18.65   
1040        88  Nations can now attack other nations with cybe...      9.40   
1132       136  Underwater filmmaker Mike deGruy has spent dec...     18.20   
1143       353  In this funny, casual talk from TEDx, writer J...     11.28   
1168       324  Colin Robertson had 3 minutes on the TED stage...      3.82   
1207        62  Taryn Simon captures the essence of vast, gene...     17.98   
1256       138  Early diagnosis of autism spectrum disorder ca...     19.73   
1406       200  Imagine you're a shipwrecked sailor adrift in ...     11.50   
1468       140  One afternoon, Kees Moeliker got a research op...     11.87   
1517       489  What is it like to raise a child who's differe...     23.45   
1549       153  Once it's created, plastic (almost) never dies...      9.33   
1623       209  For the young and unemployed in the world's bi...      9.02   
1634       217  Here's a stat worth knowing: In the UK, 63% of...     10.05   
1650        99  Oklahoma City is a midsized town that had a bi...     15.25   
1761       173  As a member of both the African American and L...     17.70   
1772       147  In 1991, Shaka Senghor shot and killed a man. ...     12.00   
1773       117  Lorrie Faith Cranor studied thousands of real ...     17.68   
1800       146  Far too many Americans are illiterate in power...     17.25   
1919        62  "We will start inhabiting outer space," says A...      7.07   
1926        49  Hooray for technology! It makes everything bet...      6.07   
1970       104  Subtle motion happens around us all the time, ...     17.95   
1978        64  The founder of Sirius XM satellite radio, Mart...     21.07   
2031        36  Dustin Yellin makes mesmerizing artwork that t...      7.53   
2158        89  Deep in the Himalayas, on the border between C...     18.90   
2227        23  TED Fellow Andrew Pelling is a biohacker, and ...      7.08   
2228        18  Chris Milk uses innovative technologies 

In [67]:
df["num_speaker"].replace(0, 1, inplace=True)
df["num_speaker"].replace(-1, 1, inplace=True)

C:\Users\saidmf\AppData\Local\Temp\ipykernel_28780\3660749878.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["num_speaker"].replace(0, 1, inplace=True)
C:\Users\saidmf\AppData\Local\Temp\ipykernel_28780\3660749878.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For exam

Finally, we're able to get back to the first issue of `main_speaker` having blanks. We can use `df['main_speaker'].isna()` to check the rows of `main_speaker` that are empty

In [68]:
df['main_speaker'].isna()

0       False
1       False
2       False
3       False
4       False
        ...  
2545    False
2546    False
2547    False
2548    False
2549    False
Name: main_speaker, Length: 2531, dtype: bool

If we look at the `description`, `main_speaker` and `name` columns for all of the blank rows, we see some useful information

In [69]:
df[df['main_speaker'].isna()][["description", "main_speaker","name"]]

description main_speaker  \
255   Nine days before TED2008, filmmaker David Hoff...          NaN   
641   Games like World of Warcraft give players the ...          NaN   
815   Join John Hardy on a tour of the Green School,...          NaN   
845   Technology is evolving us, says Amber Case, as...          NaN   
891   Lighting architect Rogier van der Heide offers...          NaN   
951   Designer Jessi Arrington packed nothing for TE...          NaN   
953   Renowned paleontologist Jack Horner has spent ...          NaN   
1250  Bartenders need to know your age, retailers ne...          NaN   
1285  Designer Jared Ficklin creates wild visualizat...          NaN   
1479  Today, thanks to better early detection, there...          NaN   
1626  Ecologist Lian Pin Koh makes a persuasive case...          NaN   
1633  The one thing all humans have in common is tha...          NaN   
1648  It's amazing what a rigged game of Monopoly ca...          NaN   
1794  Tour the deep dark world of the East German st...          NaN   
1973  Singer-songwriter Dawn Landes tells the story ...          NaN   
1998  Decades ago, few pediatricians had heard of au...          NaN   
2133  How much do you know about intellectual disabi...          NaN   
2185  "Copy editing for The New Yorker is like playi...          NaN   
2196  Secrets, disease and beauty are all written in...          NaN   

                                                   name  
255   David Hoffman: What happens when you lose ever...  
641      Jane McGonigal: Gaming can make a better world  
815                   John Hardy: My green school dream  
845                  Amber Case: We are all cyborgs now  
891      Rogier van der Heide: Why light needs darkness  
951                Jessi Arrington: Wearing nothing new  
953     Jack Horner: Building a dinosaur from a chicken  
1250      David Birch: A new way to stop identity theft  
1285  Jared Ficklin: New ways to see music (with col...  
1479  Thomas Insel: Toward a new understanding of me...  
1626   Lian Pin Koh: A drone's-eye view of conservation  
1633  David Steindl-Rast: Want to be happy? Be grateful  
1648               Paul Piff: Does money make you mean?  
1794  Hubertus Knabe: The dark secrets of a surveill...  
1973  Dawn Landes: A song for my hero, the woman who...  
1998   Steve Silberman: The forgotten history of autism  
2133  Matthew Williams: Special Olympics let me be m...  
2185  Mary Norris: The nit-picking glory of The New ...  
2196  Riccardo Sabatini: How to read the genome and ...

Without that useful information, we can always just replace empty values with a default value like so:

In [70]:
df.fillna("unknown", inplace=True)

In [71]:
df

comments                                        description  duration  \
0         4553  Sir Ken Robinson makes an entertaining and pro...     19.40   
1          265  With the same humor and humanity he exuded in ...     16.28   
2          124  New York Times columnist David Pogue takes aim...     21.43   
3          200  In an emotionally charged talk, MacArthur-winn...     18.60   
4          593  You've never seen data presented like this. Wi...     19.83   
...        ...                                                ...       ...   
2545        17  Between 2008 and 2016, the United States depor...      7.93   
2546         6  How can you study Mars without a spaceship? He...      4.83   
2547        10  Science fiction visions of the future show us ...     10.85   
2548        32  In an unmissable talk about race and politics ...     18.33   
2549         8  With more than half of the world population li...      8.65   

             event   film_date          film_date_hr  languages  \
0          TED2006  1140825600  02/24/2006, 19:00:00         60   
1          TED2006  1140825600  02/24/2006, 19:00:00         43   
2          TED2006  1140739200  02/23/2006, 19:00:00         26   
3          TED2006  1140912000  02/25/2006, 19:00:00         35   
4          TED2006  1140566400  02/21/2006, 19:00:00         48   
...            ...         ...                   ...        ...   
2545       TED2017  1496707200  06/05/2017, 20:00:00          4   
2546       TED2017  1492992000  04/23/2017, 20:00:00          3   
2547       TED2017  1492992000  04/23/2017, 20:00:00          1   
2548  TEDxMileHigh  1499472000  07/07/2017, 20:00:00          1   
2549       TED2017  1492992000  04/23/2017, 20:00:00          1   

             main_speaker                                               name  \
0            Ken Robinson          Ken Robinson: Do schools kill creativity?   
1                 Al Gore               Al Gore: Averting the climate crisis   
2             David Pogue                      David Pogue: Simplicity sells   
3           Majora Carter                 Majora Carter: Greening the ghetto   
4            Hans Rosling      Hans Rosling: The best stats you've ever seen   
...                   ...                                                ...   
2545     Duarte Geraldino  Duarte Geraldino: What we're missing in the de...   
2546  Armando Azua-Bustos  Armando Azua-Bustos: The most Martian place on...   
2547       Radhika Nagpal  Radhika Nagpal: What intelligent machines can ...   
2548     Theo E.J. Wilson  Theo E.J. Wilson: A black man goes undercover ...   
2549    Karoliina Korppoo  Karoliina Korppoo: How a video game might help...   

      num_speaker  published_date     published_date_hr  \
0               1      1151367060  06/26/2006, 20:11:00   
1               1      1151367060  06/26/2006, 20:11:00   
2               1      1151367060  06/26/2006, 20:11:00   
3               1      1151367060  06/26/2006, 20:11:00   
4               1      1151440680  06/27/2006, 16:38:00   
...           ...             ...                   ...   
2545            1      1505851216  09/19/2017, 16:00:16   
2546            1      1505919737  09/20/2017, 11:02:17   
2547            1      1506006095  09/21/2017, 11:01:35   
2548            1      1506024042  09/21/2017, 16:00:42   
2549            1      1506092422  09/22/2017, 11:00:22   

                                                ratings  \
0     [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1     [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2     [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3     [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4     [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   
...                                                 ...   
2545  [{'id': 3, 'name': 'Courageous', 'count': 24},...   
2546  [{'id': 22, 'name': 'Fascinating', 'count': 32...   
2547  [{'id': 1, 'name': 'Beautiful', 'count': 14},

In [72]:
df[df['main_speaker'] == 'unknown'].shape

(42, 19)

But here, `name` also contains the speaker's name. So we can rip it and then adjust `main_speaker` and `name` accordingly

In [73]:
df['name'].head()

0        Ken Robinson: Do schools kill creativity?
1             Al Gore: Averting the climate crisis
2                    David Pogue: Simplicity sells
3               Majora Carter: Greening the ghetto
4    Hans Rosling: The best stats you've ever seen
Name: name, dtype: object

In [74]:
df['main_speaker'] = df['name'].str.split(':').str[0]
df['main_speaker']

0              Ken Robinson
1                   Al Gore
2               David Pogue
3             Majora Carter
4              Hans Rosling
               ...         
2545       Duarte Geraldino
2546    Armando Azua-Bustos
2547         Radhika Nagpal
2548       Theo E.J. Wilson
2549      Karoliina Korppoo
Name: main_speaker, Length: 2531, dtype: object

In [75]:
df['name'] = df['name'].str.split(':').str[1]
df['name']

0                             Do schools kill creativity?
1                             Averting the climate crisis
2                                        Simplicity sells
3                                     Greening the ghetto
4                         The best stats you've ever seen
                              ...                        
2545     What we're missing in the debate about immigr...
2546                      The most Martian place on Earth
2547     What intelligent machines can learn from a sc...
2548         A black man goes undercover in the alt-right
2549     How a video game might help us build better c...
Name: name, Length: 2531, dtype: object

In [76]:
df

comments                                        description  duration  \
0         4553  Sir Ken Robinson makes an entertaining and pro...     19.40   
1          265  With the same humor and humanity he exuded in ...     16.28   
2          124  New York Times columnist David Pogue takes aim...     21.43   
3          200  In an emotionally charged talk, MacArthur-winn...     18.60   
4          593  You've never seen data presented like this. Wi...     19.83   
...        ...                                                ...       ...   
2545        17  Between 2008 and 2016, the United States depor...      7.93   
2546         6  How can you study Mars without a spaceship? He...      4.83   
2547        10  Science fiction visions of the future show us ...     10.85   
2548        32  In an unmissable talk about race and politics ...     18.33   
2549         8  With more than half of the world population li...      8.65   

             event   film_date          film_date_hr  languages  \
0          TED2006  1140825600  02/24/2006, 19:00:00         60   
1          TED2006  1140825600  02/24/2006, 19:00:00         43   
2          TED2006  1140739200  02/23/2006, 19:00:00         26   
3          TED2006  1140912000  02/25/2006, 19:00:00         35   
4          TED2006  1140566400  02/21/2006, 19:00:00         48   
...            ...         ...                   ...        ...   
2545       TED2017  1496707200  06/05/2017, 20:00:00          4   
2546       TED2017  1492992000  04/23/2017, 20:00:00          3   
2547       TED2017  1492992000  04/23/2017, 20:00:00          1   
2548  TEDxMileHigh  1499472000  07/07/2017, 20:00:00          1   
2549       TED2017  1492992000  04/23/2017, 20:00:00          1   

             main_speaker                                               name  \
0            Ken Robinson                        Do schools kill creativity?   
1                 Al Gore                        Averting the climate crisis   
2             David Pogue                                   Simplicity sells   
3           Majora Carter                                Greening the ghetto   
4            Hans Rosling                    The best stats you've ever seen   
...                   ...                                                ...   
2545     Duarte Geraldino   What we're missing in the debate about immigr...   
2546  Armando Azua-Bustos                    The most Martian place on Earth   
2547       Radhika Nagpal   What intelligent machines can learn from a sc...   
2548     Theo E.J. Wilson       A black man goes undercover in the alt-right   
2549    Karoliina Korppoo   How a video game might help us build better c...   

      num_speaker  published_date     published_date_hr  \
0               1      1151367060  06/26/2006, 20:11:00   
1               1      1151367060  06/26/2006, 20:11:00   
2               1      1151367060  06/26/2006, 20:11:00   
3               1      1151367060  06/26/2006, 20:11:00   
4               1      1151440680  06/27/2006, 16:38:00   
...           ...             ...                   ...   
2545            1      1505851216  09/19/2017, 16:00:16   
2546            1      1505919737  09/20/2017, 11:02:17   
2547            1      1506006095  09/21/2017, 11:01:35   
2548            1      1506024042  09/21/2017, 16:00:42   
2549            1      1506092422  09/22/2017, 11:00:22   

                                                ratings  \
0     [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1     [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2     [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3     [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4     [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   
...                                                 ...   
2545  [{'id': 3, 'name': 'Courageous', 'count': 24},...   
2546  [{'id': 22, 'name': 'Fascinating', 'count': 32...   
2547  [{'id': 1, 'name': 'Beautiful', 'count': 14},

Finally, let's explore how we can apply one-hot-encoding and dummy encoding on our sample csv file (which is a sample of )

In [77]:
from sklearn.preprocessing import OneHotEncoder

In [78]:
df = pd.read_csv("sample.csv")

cat_features = ["type"]                              
num_features = ["amount", "oldbalanceOrg", "newbalanceOrig", "oldbalanceDest", "newbalanceDest"]    

X_cat = df[cat_features]
X_num = df[num_features]

X_cat.head()

type
0  CASH_OUT
1   CASH_IN
2   CASH_IN
3  CASH_OUT
4  CASH_OUT

In [79]:
X_num.head()

amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  newbalanceDest
0  177585.90       25115.00            0.00       266286.98       443872.89
1  295468.97     6449944.27      6745413.24       351453.59        55984.62
2  212555.73     4543422.46      4755978.20     10013669.10      9801113.37
3    1226.94           0.00            0.00       280914.86       282141.80
4   61152.99           0.00            0.00       275058.27       336211.26

Full one-hot encoding on categorical dataset

In [80]:
ohe = OneHotEncoder()
X_cat_full = ohe.fit_transform(X_cat).toarray()

X_cat_full

array([[0., 1., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       ...,
       [0., 0., 0., 1., 0.],
       [1., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0.]], shape=(5000, 5))

In [81]:
# we can also get our new column names
ohe.get_feature_names_out(['type'])

array(['type_CASH_IN', 'type_CASH_OUT', 'type_DEBIT', 'type_PAYMENT',
       'type_TRANSFER'], dtype=object)

Notice that this a matrix of 1's and 0's that correspond to our categories in `type`. Next, we need to re-introduce these values back into our dataframe.

In [82]:
cat_names = ohe.get_feature_names_out(['type'])

encoded_df = pd.DataFrame(X_cat_full, columns=cat_names, index=df.index)

encoded_df.head()

type_CASH_IN  type_CASH_OUT  type_DEBIT  type_PAYMENT  type_TRANSFER
0           0.0            1.0         0.0           0.0            0.0
1           1.0            0.0         0.0           0.0            0.0
2           1.0            0.0         0.0           0.0            0.0
3           0.0            1.0         0.0           0.0            0.0
4           0.0            1.0         0.0           0.0            0.0

Next, we concatenate this dataframe back to our original dataframe.

In [83]:
full_df = pd.concat([X_num, encoded_df], axis=1)

full_df

amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
0     177585.90       25115.00            0.00       266286.98   
1     295468.97     6449944.27      6745413.24       351453.59   
2     212555.73     4543422.46      4755978.20     10013669.10   
3       1226.94           0.00            0.00       280914.86   
4      61152.99           0.00            0.00       275058.27   
...         ...            ...             ...             ...   
4995   47696.88         612.00        48308.88      7776230.35   
4996   10359.55        5037.00            0.00            0.00   
4997   36038.51           0.00            0.00            0.00   
4998   50598.77      131821.00       182419.77       270629.95   
4999    3113.26           0.00            0.00            0.00   

      newbalanceDest  type_CASH_IN  type_CASH_OUT  type_DEBIT  type_PAYMENT  \
0          443872.89           0.0            1.0         0.0           0.0   
1           55984.62           1.0            0.0         0.0           0.0   
2         9801113.37           1.0            0.0         0.0           0.0   
3          282141.80           0.0            1.0         0.0           0.0   
4          336211.26           0.0            1.0         0.0           0.0   
...              ...           ...            ...         ...           ...   
4995      7728533.47           1.0            0.0         0.0           0.0   
4996            0.00           0.0            0.0         0.0           1.0   
4997            0.00           0.0            0.0         0.0           1.0   
4998       220031.18           1.0            0.0         0.0           0.0   
4999            0.00           0.0            0.0         0.0           1.0   

      type_TRANSFER  
0               0.0  
1               0.0  
2               0.0  
3               0.0  
4               0.0  
...             ...  
4995            0.0  
4996            0.0  
4997            0.0  
4998            0.0  
4999            0.0  

[5000 rows x 10 columns]

To implement dummy-encoding, we simply nee dinclude `drop='first'`

In [84]:
ohe = OneHotEncoder(drop='first')
X_cat_full = ohe.fit_transform(X_cat).toarray()

X_cat_full

array([[1., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       ...,
       [0., 0., 1., 0.],
       [0., 0., 0., 0.],
       [0., 0., 1., 0.]], shape=(5000, 4))